pkg/controller/direct/compute/firewallpolicyrule/firewallpolicyrule_controller.go

```diff
@@ -0,0 +1,290 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package firewallpolicyrule
+
+import (
+	"context"
+	"fmt"
+
+	gcp "cloud.google.com/go/compute/apiv1"
+	computepb "cloud.google.com/go/compute/apiv1/computepb"
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/compute/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const ctrlName = "firewallpolicyrule-controller"
+
+func init() {
+	registry.RegisterModel(krm.ComputeFirewallPolicyRuleGVK, NewFirewallPolicyRuleModel)
+}
+
+func NewFirewallPolicyRuleModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &firewallPolicyRuleModel{config: config}, nil
+}
+
+type firewallPolicyRuleModel struct {
+	config *config.ControllerConfig
+}
+
+// model implements the Model interface.
+var _ directbase.Model = &firewallPolicyRuleModel{}
+
+type firewallPolicyRuleAdapter struct {
+	firewallPolicy         string
+	priority               int64
+	firewallPoliciesClient *gcp.FirewallPoliciesClient
+	desired                *krm.ComputeFirewallPolicyRule
+	actual                 *computepb.FirewallPolicyRule
+	reader                 client.Reader
+}
+
+var _ directbase.Adapter = &firewallPolicyRuleAdapter{}
+
+func (m *firewallPolicyRuleModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.ComputeFirewallPolicyRule{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Set label managed-by-cnrm: true
+	obj.ObjectMeta.Labels["managed-by-cnrm"] = "true"
+
+	// Get firewall policy
+	firewallPolicyRef, err := ResolveComputeFirewallPolicy(ctx, reader, obj, obj.Spec.FirewallPolicyRef)
+	if err != nil {
+		return nil, err
+
+	}
+	obj.Spec.FirewallPolicyRef.External = firewallPolicyRef.External
+	firewallPolicy := obj.Spec.FirewallPolicyRef.External
+
+	// Get priority
+	priority := obj.Spec.Priority
```

Comment:
What is priority? This seems to be the firewall policy rule's ID.

Reply:
I do not think we can use priority itself as **unique** ID. Different rule with same priority can be associated with different policy, i.e.

policy1: {rule1(priority 100), rule2(priority 300)}
policy2: {rule3(priority 100), rule4(priority 300)}

If we include the policy i.e. `policies/policyName/priorities/priorityNumber` that could be used as unique identifier. It's still a bit "unusual" because this is not an expression exist in GCP. The GCP request to get a rule looks like:` GET https://compute.googleapis.com/compute/v1/locations/global/firewallPolicies/${firewallPolicyID}/getRule?priority=9000`, not `GET https://compute.googleapis.com/compute/v1/locations/global/firewallPolicies/${firewallPolicyID}/priorities/9000`

Reply:
Added externalRef in #3035